Convert to using UV for environment management

angus-g opened the pull request:
Gives us a more reliable and easy to update `uv.lock` file, hopefully also simplifies both the development and build workflows.

dsroberts commented:
I'm a little confused as to how `uv.lock` is meant to work. It doesn't seem to be respected at all. Running `uv run -- mkdocs serve` forces `uv.lock` to be updated, and running `uv run --locked -- mkdocs serve` causes `uv` to exit with an error. My concern is that every different author will wind up committing a different `uv.lock` to the repository for every change made, leading to a pile of useless changes going into the repo. Maybe I'm misunderstanding how its meant to work?

dsroberts commented:
This is less than ideal
```
[dale@DESKTOP-93BIFKT g-adopt.github.io]$ git diff
[dale@DESKTOP-93BIFKT g-adopt.github.io]$ uv run -- mkdocs serve
Using CPython 3.11.9 interpreter at: /usr/bin/python3.11
Creating virtual environment at: .venv
Installed 83 packages in 133ms
INFO    -  DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
           given by the platformdirs library.  To remove this warning and
           see the appropriate new directories, set the environment variable
           `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
           The use of platformdirs will be the default in `jupyter_core` v6
             File
           "/home/dale/g-adopt.github.io/.venv/lib64/python3.11/site-packages/jupyter_core/utils/__init__.py",
           line 90, in deprecation
               warnings.warn(message, DeprecationWarning, stacklevel=stacklevel + 1)
             File
           "/home/dale/g-adopt.github.io/.venv/lib64/python3.11/site-packages/jupyter_client/connect.py", line
           22, in
               from jupyter_core.paths import jupyter_data_dir, jupyter_runtime_dir, secure_write
INFO    -  Building documentation...
INFO    -  FutureWarning: The 'import' key is renamed 'inventories' for the Python handler
             File
           "/home/dale/g-adopt.github.io/.venv/lib64/python3.11/site-packages/mkdocstrings/handlers/base.py",
           line 682, in get_handler
               self._handlers[name] = module.get_handler(**kwargs)
             File
           "/home/dale/g-adopt.github.io/.venv/lib64/python3.11/site-packages/mkdocstrings_handlers/python/handler.py",
           line 370, in get_handler
               warn("The 'import' key is renamed 'inventories' for the Python handler", FutureWarning,
           stacklevel=1)
INFO    -  Cleaning site directory
INFO    -  Documentation built in 0.41 seconds
INFO    -  [17:10:25] Watching paths for changes: 'docs', 'mkdocs.yml'
INFO    -  [17:10:25] Serving on http://127.0.0.1:8000/
^CINFO    -  Shutting down...
[dale@DESKTOP-93BIFKT g-adopt.github.io]$ git diff --name-only
uv.lock
[dale@DESKTOP-93BIFKT g-adopt.github.io]$ git diff | wc -l
313
```

angus-g commented:
Maybe I need to be stricter on Python versions? My machine used Python 3.13, and it looks like you're on 3.11, so I guess it's changing versions because of that. I will freely admit I haven't used uv in this context yet!

dsroberts commented:
The weird thing is that all its upped the minimum version to 3.11 and removed all of the 3.10 packages from `uv.lock`. `pyproject.toml` says `requires-python = ">=3.11"` which would seem to contradict what was in the version of `uv.lock` committed to the repo.

dsroberts commented on pyproject.toml:
```diff
@@ -0,0 +1,15 @@
+version = "0.1.0"
+description = "The website and documentation for the g-adopt project"
+readme = "README.md"
+requires-python = ">=3.11"
```
```suggestion
requires-python = ">=3.10"
```
I've confirmed this builds and runs fine on Python 3.10, and this change fixes the `uv.lock` changes problem. I think I see the value in `uv.lock` now, having full set of dependencies anyone could have all there makes it easier to find what might have caused a breaking change.

angus-g replied:
Curious, I was hitting a weird incompatibility with 3.10 which is why I bumped the version in the first place...

dsroberts replied:
How strange...
```
[dale@DESKTOP-93BIFKT g-adopt.github.io]$ uv run -- mkdocs serve
Using CPython 3.10.16 interpreter at: /opt/python3/3.10.16/bin/python3
Creating virtual environment at: .venv
Installed 85 packages in 160ms
INFO    -  DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
           given by the platformdirs library.  To remove this warning and
           see the appropriate new directories, set the environment variable
           `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
           The use of platformdirs will be the default in `jupyter_core` v6
             File
           "/home/dale/g-adopt.github.io/.venv/lib/python3.10/site-packages/jupyter_core/utils/__init__.py",
           line 90, in deprecation
               warnings.warn(message, DeprecationWarning, stacklevel=stacklevel + 1)
             File "/home/dale/g-adopt.github.io/.venv/lib/python3.10/site-packages/jupyter_client/connect.py",
           line 22, in
               from jupyter_core.paths import jupyter_data_dir, jupyter_runtime_dir, secure_write
INFO    -  Building documentation...
INFO    -  FutureWarning: The 'import' key is renamed 'inventories' for the Python handler
             File
           "/home/dale/g-adopt.github.io/.venv/lib/python3.10/site-packages/mkdocstrings/handlers/base.py",
           line 682, in get_handler
               self._handlers[name] = module.get_handler(**kwargs)
             File
           "/home/dale/g-adopt.github.io/.venv/lib/python3.10/site-packages/mkdocstrings_handlers/python/handler.py",
           line 370, in get_handler
               warn("The 'import' key is renamed 'inventories' for the Python handler", FutureWarning,
           stacklevel=1)
INFO    -  Cleaning site directory
INFO    -  Documentation built in 0.46 seconds
INFO    -  [18:03:36] Watching paths for changes: 'docs', 'mkdocs.yml'
INFO    -  [18:03:36] Serving on http://127.0.0.1:8000/
```

angus-g replied:
:exploding_head:

angus-g replied:
I think my computer is just silly.

dsroberts reviewed (commented):
Seems like a reasonable change, makes for much simpler test environment setup.

angus-g commented:
> The weird thing is that all its upped the minimum version to 3.11 and removed all of the 3.10 packages from `uv.lock`. `pyproject.toml` says `requires-python = ">=3.11"` which would seem to contradict what was in the version of `uv.lock` committed to the repo. ETA: Yeah, this appears to be the problem, I changed `pyproject.toml` to `requires-python = ">=3.10"` and it didn't change `uv.lock` on running `uv run -- mkdocs serve`.

Yep, I just completely forgot to add the updated uv.lock after bumping the minimum version... My bad! Hopefully it's fine now.

angus-g commented:
I guess we'll need this, or some other requirements change to fix the failing builds: https://github.com/g-adopt/g-adopt.github.io/actions/runs/13534994140

dsroberts commented:
https://github.com/mkdocstrings/autorefs/issues/68

angus-g commented:
Build passes (using Python 3.13 from the Firedrake container) and also runs locally with both Python 3.10 and 3.13.

dsroberts reviewed (approved):
LGTM
